### PR TITLE
Cpp spatial weights

### LINKED ISSUE #65: cEpiabm - Spatial weightings
Replicate python code in C++ - a lot of the framework should be there already! P.S make sure to branch off Nic's PR #36 to have the up to date version

### PR DESCRIPTION
**Pull request by Elizabeth-Hayman**
Spatial sweep with distance weights
 Closes #65

### CODE REVIEW

**Comment by KCGallagher** [comment]
@Elizabeth-Hayman is this still a draft or are you wanting a review? It looks like this has branched off a more recent version of cEpiabm than the version on main, so may want to wait until that is merged in before looking at this. Also there are a lot of conflicts so probably worth resolving those before any reviews

File: cEpiabm/src/sweeps/household_sweep.cpp
Comment (KCGallagher) [inline cEpiabm/src/sweeps/household_sweep.cpp:84]: Think you have accidentally merged in an old version of the C++ - @NicholasFan235 changed this the other way recently? Maybe worth checking no other changes have been reversed?

File: cEpiabm/src/sweeps/spatial_sweep.cpp
Comment (KCGallagher) [inline cEpiabm/src/sweeps/spatial_sweep.cpp:45]: Query - is `cpp-dev` a separate branch? Does that mean we need that branch in for this to work?

File: cEpiabm/src/utility/distance_metrics.cpp
Comment (KCGallagher) [inline cEpiabm/src/utility/distance_metrics.cpp:24]: Is this a placeholder?

File: cEpiabm/test/utility/test_distance_metrics.cpp
Comment (KCGallagher) [inline cEpiabm/test/utility/test_distance_metrics.cpp:19]:
```suggestion
    auto loc2 = std::make_pair(-1.0, 2.0);
```
Tiny thing but if we have two distances it might be good for them to have x and y separation to check this is initialised properly (and crossing the origin for the same reason). I know we aren't testing to the same level on the C++ but thought I'd suggest this as it seems an easy change to make! Ditto on the test cases below

**Review by KCGallagher — CHANGES_REQUESTED** [review]
I wasn't sure if this was ready for review, but had some time so thought I would have a look - please ignore comments here if it is still WIP though! 
Overall it looks really great - I added a few minor comments/questions but nothing that would take more than 5 minutes to go through!

One thing that is slightly concerning - @NicholasFan235 this seems to reverse your change on the factor of 0.5 in the FOI? I have flagged it, and can't see anything else, but maybe worth going through to check no other recent changes have been overwritten

File: cEpiabm/src/utility/distance_metrics.cpp
Comment (Elizabeth-Hayman) [reply cEpiabm/src/utility/distance_metrics.cpp:24]: Very much so

File: cEpiabm/src/sweeps/spatial_sweep.cpp
Comment (Elizabeth-Hayman) [reply cEpiabm/src/sweeps/spatial_sweep.cpp:45]: doCovidsim is a true/false flag that for a simulation needs to be initialised in a config file somewhere. But everything will work with the tests just hardcoding the flag.

File: cEpiabm/src/sweeps/household_sweep.cpp
Comment (Elizabeth-Hayman) [reply cEpiabm/src/sweeps/household_sweep.cpp:84]: so we DON'T want the factor of 0.5?

Comment (KCGallagher) [reply cEpiabm/src/sweeps/household_sweep.cpp:84]: No, but I think it must have been added by merging back in an old branch so worth checking with @NicholasFan235  whether any other changes have been overwritten too

File: cEpiabm/src/sweeps/spatial_sweep.cpp
Comment (KCGallagher) [reply cEpiabm/src/sweeps/spatial_sweep.cpp:45]: Oh that makes sense - would the config file add it to a global namespace or would it need to be read in from parameters here?

File: cEpiabm/src/utility/distance_metrics.cpp
Comment (KCGallagher) [reply cEpiabm/src/utility/distance_metrics.cpp:24]: How easy would it be to add the covidsim implementation, so we can compare to them directly?

File: cEpiabm/src/sweeps/spatial_sweep.cpp
Comment (NicholasFan235) [reply cEpiabm/src/sweeps/spatial_sweep.cpp:45]: Fixed this in the cpp-dev branch, happy for this to go into main first and then we can pull in cpp-dev

File: cEpiabm/src/utility/distance_metrics.cpp
Comment (KCGallagher) [reply cEpiabm/src/utility/distance_metrics.cpp:24]: Okay we can get this in but given this is an unmarked placeholder I will also make an issue and allocate it to you so we don't forget!

**Review by KCGallagher — APPROVED** [review]

**Review by rccreswell — APPROVED** [review]

**Pull request merged**
